src/lib.rs

```diff
@@ -101,9 +116,58 @@ static MIN_TIME_PERIOD: once_cell::sync::Lazy<winapi::shared::minwindef::UINT> =
 pub(crate) fn thread_sleep(duration: Duration) {
     unsafe {
         use winapi::um::timeapi::{timeBeginPeriod, timeEndPeriod};
+
+        /*
+        This is a *BIG* no no..  The problem here is that timeBeginPeriod has
+        global impact on the entire OS.  The overhead of changing this value
+        regularly is ... not good, since all processes (not just threads) in the
+        system will now have different accuracy for sleeps, system interrupt
+        rates are increased, etc etc.  Not to mention that the cost of the pair
+        of calls is exceptionally high.  From the docs:
+
+        "Use caution when calling timeBeginPeriod, as frequent calls can
+        significantly affect the system clock, system power usage, and the scheduler."
+
+        To keep using this, only do it once a frame at most and even that is
+        not really a great idea since it can trigger a number of bugs in the
+        overall system where timers change accuracy unexpectedly in completely
+        unrelated processes.
```

Comment:
I think that there's an implicit high performance use case for spin_sleep. My thinking here is; if you are prepared to actually busy wait a thread, I can't imagine that you are _not_ prepared to use the OS's highest accuracy (but still quite poor) sleep mode.

Reply:
So, the problem with this API is that it mucks with global OS settings which are setup at Windows startup too get the best balance of performance to overhead for the CPU+Memory of the system.  Setting it faster like this causes a very notable set of issues for the OS in terms that it starts servicing things at a higher rate, decreases thread time slices etc, all of which cranks up general OS overhead such that both overall performance drops notably and for laptops you drain the batteries quite significantly faster.
 
The last time I specifically tested this was on Win7 for Swtor and ME3 when I was fixing a timer system for those games, so it is possible though unlikely that Win10 has solved some of the issue.  But, at the time the overall OS overhead jumped up about 5% (i.e. every thread was 5% slower) and the battery life of laptops dropped by over 20%.  Additionally, at that time, the call to begin/end period itself was often in the 15k cycle range, meaning it was often longer than the desired delay before we even called sleep, so we were actually making things worse.

Performance aside, as the doc for the API mentions, modifying these settings will cause clock drift and can break other things on the system.  A very specific case from Swtor was that it caused us a lot of sound system problems because the timers for feeding sound buffers became just a little wrong, which when streaming that much VOX, ambient music/noise and of course iconic Star Wars music was very time sensitive.  After removing use of the API the "crackling" audio bug went away, go figure.

All said and done, I'm interested if some of these issues have gone away but not very hopeful that they have been fixed since this is something MS suggests not using at all.  The example case is that the XBox 360 and the XBone SDK's don't have this API at all.  Of course on those platforms, sleep has 8-10 ms accuracy so it's not actually needed.